File: back/src/main/java/com/triptrace/global/globalExceptionHandler/GlobalExceptionHandler.java
package com.triptrace.global.globalExceptionHandler;

import static org.springframework.http.HttpStatus.*;

import java.util.Comparator;
import java.util.NoSuchElementException;
import java.util.stream.Collectors;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.MissingRequestHeaderException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.multipart.MaxUploadSizeExceededException;

import com.triptrace.domain.image.image.processing.exception.ImageProcessException;
import com.triptrace.global.error.DefaultErrorCode;
import com.triptrace.global.exception.ServiceException;
import com.triptrace.global.rsData.RsData;

import jakarta.validation.ConstraintViolationException;
import lombok.RequiredArgsConstructor;

@RestControllerAdvice
@RequiredArgsConstructor
public class GlobalExceptionHandler {
    @ExceptionHandler(NoSuchElementException.class)
    public ResponseEntity<RsData<Void>> handle(NoSuchElementException ex) {
        return new ResponseEntity<>(
            new RsData<>(
                DefaultErrorCode.NOT_FOUND.getCode(),
                DefaultErrorCode.NOT_FOUND.getMessage()
            ),
            NOT_FOUND
        );
    }

    @ExceptionHandler(ConstraintViolationException.class)
    public ResponseEntity<RsData<Void>> handle(ConstraintViolationException ex) {
        String message = ex.getConstraintViolations()
            .stream()
            .map(violation -> {
                String field = violation.getPropertyPath().toString().split("\\.", 2)[1];
                String[] messageTemplateBits = violation.getMessageTemplate()
                    .split("\\.");
                String code = messageTemplateBits[messageTemplateBits.length - 2];
                String _message = violation.getMessage();

                return "%s-%s-%s".formatted(field, code, _message);
            })
            .sorted(Comparator.comparing(String::toString))
            .collect(Collectors.joining("\n"));

        return new ResponseEntity<>(
            new RsData<>(
                DefaultErrorCode.BAD_REQUEST.getCode(),
                message
            ),
            BAD_REQUEST
        );
    }

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<RsData<Void>> handle(MethodArgumentNotValidException ex) {
        String msg = ex.getBindingResult()
            .getAllErrors()
            .stream()
            .filter(error -> error instanceof FieldError)
            .map(error -> (FieldError)error)
            .map(error -> error.getField() + "-" + error.getCode() + "-" + error.getDefaultMessage())
            .sorted()
            .collect(Collectors.joining("\n"));

        return new ResponseEntity<>(
            new RsData<>(
                DefaultErrorCode.BAD_REQUEST.getCode(),
                msg
            ),
            BAD_REQUEST
        );
    }

    @ExceptionHandler(HttpMessageNotReadableException.class)
    public ResponseEntity<RsData<Void>> handle(HttpMessageNotReadableException ex) {
        return new ResponseEntity<>(
            new RsData<>(
                DefaultErrorCode.BAD_REQUEST.getCode(),
                DefaultErrorCode.BAD_REQUEST.getMessage()
            ),
            BAD_REQUEST
        );
    }

    @ExceptionHandler(MissingRequestHeaderException.class)
    public ResponseEntity<RsData<Void>> handle(MissingRequestHeaderException ex) {
        return new ResponseEntity<>(
            new RsData<>(
                DefaultErrorCode.BAD_REQUEST.getCode(),
                "%s-%s-%s".formatted(
                    ex.getHeaderName(),
                    "NotBlank",
                    ex.getLocalizedMessage()
                )
            ),
            BAD_REQUEST
        );
    }

    @ExceptionHandler(ServiceException.class)
    public ResponseEntity<RsData<Void>> handle(ServiceException ex) {
        RsData<Void> rsData = ex.getRsData();

        return new ResponseEntity<>(
            rsData,
            ResponseEntity
                .status(rsData.statusCode())
                .build()
                .getStatusCode()
        );
    }

    @ExceptionHandler(ImageProcessException.class)
    public ResponseEntity<RsData<Void>> handle(ImageProcessException ex) {
        RsData<Void> rsData = new RsData<>(
            ex.getResultCode(),
            ex.getMsg(),
            null
        );
        return new ResponseEntity<>(
            rsData,
            ResponseEntity
                .status(rsData.statusCode())
                .build()
                .getStatusCode()
        );
    }

    @ExceptionHandler(MaxUploadSizeExceededException.class)
    public ResponseEntity<RsData<Void>> handle(MaxUploadSizeExceededException ex) {
        return new ResponseEntity<>(
            new RsData<>(
                DefaultErrorCode.PAYLOAD_TOO_LARGE.getCode(),
                "%s-%s".formatted(
                    DefaultErrorCode.PAYLOAD_TOO_LARGE.getMessage(),
                    ex.getLocalizedMessage()
                )
            ),
            HttpStatus.PAYLOAD_TOO_LARGE
        );
    }
}

File: back/src/main/java/com/triptrace/global/globalExceptionHandler/GlobalExceptionHandler.kt
package com.triptrace.global.globalExceptionHandler

import com.triptrace.domain.image.image.exception.ImageProcessException
import com.triptrace.global.error.DefaultErrorCode
import com.triptrace.global.exception.ServiceException
import com.triptrace.global.rsData.RsData
import jakarta.validation.ConstraintViolationException
import org.springframework.http.HttpStatus
import org.springframework.http.HttpStatusCode
import org.springframework.http.ResponseEntity
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.validation.FieldError
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.MissingRequestHeaderException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.multipart.MaxUploadSizeExceededException
import java.util.NoSuchElementException

@RestControllerAdvice
class GlobalExceptionHandler {

    @ExceptionHandler(NoSuchElementException::class)
    fun handle(ex: NoSuchElementException): ResponseEntity<RsData<Void>> {
        return ResponseEntity(
            RsData(
                DefaultErrorCode.NOT_FOUND.getCode(),
                DefaultErrorCode.NOT_FOUND.getMessage()
            ),
            HttpStatus.NOT_FOUND
        )
    }

    @ExceptionHandler(ConstraintViolationException::class)
    fun handle(ex: ConstraintViolationException): ResponseEntity<RsData<Void>> {
        val message = ex.constraintViolations
            .map { violation ->
                val field = violation.propertyPath.toString().split(".", limit = 2)[1]
                val messageTemplateBits = violation.messageTemplate.split(".")
                val code = messageTemplateBits[messageTemplateBits.size - 2]

                "$field-$code-${violation.message}"
            }
            .sorted()
            .joinToString("\n")

        return ResponseEntity(
            RsData(
                DefaultErrorCode.BAD_REQUEST.getCode(),
                message
            ),
            HttpStatus.BAD_REQUEST
        )
    }

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handle(ex: MethodArgumentNotValidException): ResponseEntity<RsData<Void>> {
        val msg = ex.bindingResult
            .allErrors
            .filterIsInstance<FieldError>()
            .map { error -> "${error.field}-${error.code}-${error.defaultMessage}" }
            .sorted()
            .joinToString("\n")

        return ResponseEntity(
            RsData(
                DefaultErrorCode.BAD_REQUEST.getCode(),
                msg
            ),
            HttpStatus.BAD_REQUEST
        )
    }

    @ExceptionHandler(HttpMessageNotReadableException::class)
    fun handle(ex: HttpMessageNotReadableException): ResponseEntity<RsData<Void>> {
        return ResponseEntity(
            RsData(
                DefaultErrorCode.BAD_REQUEST.getCode(),
                DefaultErrorCode.BAD_REQUEST.getMessage()
            ),
            HttpStatus.BAD_REQUEST
        )
    }

    @ExceptionHandler(MissingRequestHeaderException::class)
    fun handle(ex: MissingRequestHeaderException): ResponseEntity<RsData<Void>> {
        return ResponseEntity(
            RsData(
                DefaultErrorCode.BAD_REQUEST.getCode(),
                "${ex.headerName}-NotBlank-${ex.localizedMessage}"
            ),
            HttpStatus.BAD_REQUEST
        )
    }

    @ExceptionHandler(ServiceException::class)
    fun handle(ex: ServiceException): ResponseEntity<RsData<Void>> {
        val rsData = ex.getRsData()

        // resultCode 앞자리에서 뽑아 둔 statusCode를 그대로 HTTP 상태로 쓴다.
        return ResponseEntity(rsData, HttpStatusCode.valueOf(rsData.statusCode))
    }

    @ExceptionHandler(ImageProcessException::class)
    fun handle(ex: ImageProcessException): ResponseEntity<RsData<Void>> {
        val rsData = RsData<Void>(
            ex.resultCode,
            ex.msg,
            null
        )

        return ResponseEntity(rsData, HttpStatusCode.valueOf(rsData.statusCode))
    }

    @ExceptionHandler(MaxUploadSizeExceededException::class)
    fun handle(ex: MaxUploadSizeExceededException): ResponseEntity<RsData<Void>> {
        return ResponseEntity(
            RsData(
                DefaultErrorCode.PAYLOAD_TOO_LARGE.getCode(),
                "${DefaultErrorCode.PAYLOAD_TOO_LARGE.getMessage()}-${ex.localizedMessage}"
            ),
            HttpStatus.PAYLOAD_TOO_LARGE
        )
    }
}
